feat: POST method at /compute to calculate bowling scores

Hevhove opened the pull request:
Completed assignment for bowling http server:

Feats:
- Implemented /compute POST endpoint, which responds with 200 OK if the request was formatted correctly, and a calculated bowling score

To test, run `npm run start` and pass the tests defined in the test folder with `npm run test`

Usage example (outputs of /tests/request.sh):
```
❯ ./tests/request.sh
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST /compute HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 96
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Type: application/json; charset=utf-8
< Content-Length: 13
< ETag: W/"d-kO0pSYv8WiV9YOMhW0fuOrD2J4Q"
< Date: Mon, 12 Jun 2023 16:44:47 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
<
* Connection #0 to host localhost left intact
{"score":300}%
```

WitoDelnat commented on src/compute.js:
```diff
@@ -0,0 +1,38 @@
+"use strict";
```
Gitignore files derived from source code. It's important to never commit them as they bloat the repo.

Hevhove replied:
Just to clarify: any transpiled JS from my TypeScript files should be in .gitignore?
I have added this script to package.json: `"clean": "find ./src -name '*.js' -type f -delete",`. But it should be remembered before committing. Is this a good practice?

WitoDelnat replied:
No, you add a _.gitignore_ file: https://git-scm.com/docs/gitignore.

Most commonly you configure TypeScript to emit the output to a `/dist` directory instead side-by-side in `src`. Afterwards you simply put `dist/**` in the _.gitignore_ file. No idea why I didn't do that here to be fair, I guess I was in a rush with the setup but you can try it yourself to learn a bit about the tsconfig.

The result is that you can simply commit everything and it will not take it into consideration. Do note that since you committed it once you will have to delete it from git as it's currently being tracked. It's a good git exercise to explore how to do that.

WitoDelnat commented on src/compute.ts:
```diff
@@ -1,5 +1,33 @@
+    // Loop over all of the frames and calculate bonusScores
+    for (let i = 0; i < game.length; i++) {
+        const frame = game[i];
+        const frameScore = game[i].reduce((a, b) => a + b, 0);
```
Use better variable names: a -> sum, b -> roll

Hevhove replied:
changed them :)

WitoDelnat commented on src/compute.ts:
```diff
@@ -1,5 +1,33 @@
+        if (frameScore === 10) {
+            // Get the next frame and the next next frame
+            // || [] is a fallback in case the next frame is undefined
+            const nextFrame = game[i + 1] || [];
```
Nitpick prefer `??` over `||`. `??` checks whether it is undefined or null while `||` checks if it is falsy, this often gives bugs when working with strings (`""`) or numbers (`0`), though here it will work because it are objects but it's a good habbit.

Hevhove replied:
Very interesting I didn't consider that. Making the change!

WitoDelnat commented on src/server.ts:
```diff
@@ -1,16 +1,59 @@
+
+    // if all of the above didn't trigger, we're good to go!
+    return true;
+}
```
Nitpick: Always add white line add end of file

Hevhove replied:
Cool, I added this setting to my vscode setup: `Files: Insert Final Newline
When enabled, insert a final new line at the end of the file when saving it.`

Hevhove commented:
I have commented my changes above @WitoDelnat. 

Additionally, I have committed 3 new test cases which are all passing correctly. I'm not too experienced with this yet, do you have any suggestions to add to get a range of test cases? Currently I have added a game of 0s, a game of spares, a random game etc. Also implemented a test case to check if we get back 400 on invalid input, but there are many variations we could test of "bad input"

WitoDelnat commented:
Normally this is more than sufficient. In the end it depends on how error prone the code is - most often I rarely write unit tests and rely more on blackbox tests (opinions vary here!).

When I wrote this task I added like 6 random games playing with the last frame just to make sure people covered all cases, but this was more exceptional.
